xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesEndpointFetchingService.java

```diff
@@ -134,15 +136,24 @@ protected void handleXdsResource(String path, String contentAsText, String group
             oldUpdater.close();
         }
         final KubernetesEndpointsUpdater updater =
-                new KubernetesEndpointsUpdater(commandExecutor, kubernetesEndpointGroup, executorService,
+                new KubernetesEndpointsUpdater(commandExecutor, future, executorService,
                                                groupName, watcherName, endpointWatcher.getClusterName());
         updaters.put(watcherName, updater);
-        kubernetesEndpointGroup.addListener(endpoints -> {
-            if (endpoints.isEmpty()) {
-                return;
+        future.handle((kubernetesEndpointGroup, cause) -> {
+            if (cause != null) {
+                // Do not remove the updater from updaters because it can remove the updater that is created
+                // by the next commit. The updater will be removed only when the file or group is removed.
+                updater.close();
+                return null;
             }
-            executorService.execute(updater::maybeSchedule);
-        }, true);
+            kubernetesEndpointGroup.addListener(endpoints -> {
+                if (endpoints.isEmpty()) {
+                    return;
+                }
+                executorService.execute(updater::maybeSchedule);
```

Comment:
How about passing `kubernetesEndpointGroup` as an argument to `maybeSchedule` to avoid calling `kubernetesEndpointGroupFuture.join()`?

Reply:
That's a good suggestion. 👍